imod/mf6/wel.py

```diff
@@ -220,31 +231,63 @@ def clip_box(
         ----------
         time_min: optional
         time_max: optional
-        z_min: optional, float
-        z_max: optional, float
+        layer_min: optional, int
+        layer_max: optional, int
         x_min: optional, float
         x_max: optional, float
         y_min: optional, float
         y_max: optional, float
+        top: optional, GridDataArray
+        bottom: optional, GridDataArray
+        state_for_boundary: optional, GridDataArray
 
         Returns
         -------
         sliced : Package
         """
+        if (layer_max or layer_min) and (top is None or bottom is None):
+            raise ValueError(
+                "When clipping by layer both the top and bottom should be defined"
+            )
+
+        if top is not None:
+            if not isinstance(top, GridDataArray) or "layer" not in top.coords:
+                top = create_layered_top(bottom, top)
 
         # The super method will select in the time dimension without issues.
         new = super().clip_box(time_min=time_min, time_max=time_max)
 
         ds = new.dataset
 
+        z_max = None if layer_max is None else top.isel(layer=layer_max)
+        z_min = None if layer_min is None else bottom.isel(layer=layer_min)
+
+        # if z is a grid select the the values at the well locations and drop the dimensions
+        if (z_max is not None) and ("x" in z_max.coords or "y" in z_max.coords):
+            z_max = z_max.sel(x=ds["x"], y=ds["y"], method="nearest").drop_vars(
+                lambda x: x.coords
+            )
+
+        if (z_min is not None) and ("x" in z_min.coords or "y" in z_min.coords):
+            z_min = z_min.sel(x=ds["x"], y=ds["y"], method="nearest").drop_vars(
+                lambda x: x.coords
+            )
+
+        if z_max is not None:
+            ds["screen_top"] = ds["screen_top"].clip(None, z_max)
+        if z_min is not None:
+            ds["screen_bottom"] = ds["screen_bottom"].clip(z_min, None)
+
```

Comment:
Already better, but there are still some issues: 

1. This doesn't work for unstructured grids, these do not have "x" and "y" in coords, so then the if statements fails, as well as the call to ``.sel``, this should be ``.ugrid.sel``. Luckily we don't have to reinvent the wheel. We already wrote ``imod.typing.grid.is_spatial_2D`` for this and ``imod.select.points_values``.
2.  ``z_min`` and ``z_max`` can be None, but can also be first a grid and then be converted to a 1d array. I think it is best to keep ``z_min``, ``z_max`` to something that is accepted by ``values_within_range``, so not as grid.

In conclusion: 

```suggestion
        z_max_grid = None if layer_max is None else top.isel(layer=layer_max)
        z_min_grid = None if layer_min is None else bottom.isel(layer=layer_min)

        # if z is a grid select the the values at the well locations and drop the dimensions
        if (z_max_grid is not None) and is_spatial_2D(z_max_grid):
            z_max =  imod.select.points_values(z_max_grid, x=ds["x"].values, y=ds["y"].values, out_of_bounds="ignore")
            )
        else: 
            z_max = None

        if ( z_min_grid is not None) and is_spatial_2D(z_min_grid):
            z_min =  imod.select.points_values(z_min_grid, x=ds["x"].values, y=ds["y"].values, out_of_bounds="ignore")
            )
        else: 
            z_min = None
```

The call to ``is_spatial_2D(z_max_grid)`` is probably overdoing things, so you can probably remove that one.

Reply:
This worked like a charm. I added 2 unit tests for unstructured grids and with your code they both pass